### PR TITLE
BytesBuf set_buffer( ) method for arbitrary binary read/write on data objects [main]

### PR DESCRIPTION
**Pull request by d-w-moore**
Includes:
- cherry-pick of https://github.com/irods/irods_rule_engine_plugin_python/pull/106
- fix for cmake allowing Python rule plugin's `main` branch to build.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -190,6 +190,35 @@ def get_irods_username (rule_args, callback, rei):
   rule_args [0] = username
 ```
 
+## Special methods
+
+Some iRODS objects used with rules and microservices have been given utility methods.
+
+An example is the map() method of a :code:`PluginContext` PEP argument::
+    def  pep_resource_resolve_hierarchy_pre(args,_,_):
```
Comment (korydraughn) [inline README.md:198]: extra space after `def`

File: README.md [hunk 2]
```diff
@@ -190,6 +190,40 @@ def get_irods_username (rule_args, callback, rei):
...
+
+In version 4.2.11.2 of the Python plugin, `BytesBuf` has methods to set and access byte content.
+These can be used to achieve true binary reads and writes, to and from data objects.
+
+For example:
+```
+    def my_rule(args, callback, rei):
+        buf = irods_types.BytesBuf()
+        buf.set_buffer( os.urandom(256) )
+        callback.msiDataObjectWrite( descriptor, buf, 0)  #-- Length of 0 writes the full buffer.
```
Comment (korydraughn) [inline README.md:212]: trailing comment is a little confusing.

i can't figure out if something is missing or if i'm just parsing it wrong

Comment (d-w-moore) [reply README.md:212]: Maybe just "write full buffer" ?

Comment (korydraughn) [reply README.md:212]: does zero instruct the msi to write the full buffer?

Comment (d-w-moore) [reply README.md:212]: Checked on it, it's an outParam ... so it doesn't affect the msi operation. It's just there for iRODS to return the # bytes actually written.

Comment (korydraughn) [reply README.md]: cool cool.

File: README.md (hunk 1 above)
Comment (d-w-moore) [reply README.md]: Done !

**Comment by d-w-moore** [comment]
squashed

**Comment by trel** [comment]
please # them

**Comment by d-w-moore** [comment]
#'d

**Pull request merged**